### ISSUE TITLE
How do you suggest to handle case when Async body writes to database?

### ISSUE THREAD

**Issue by ursusursus** [airbnb/MvRx#346]
Hi, I'd like some advice on this common case.

Say we have a `Product(id: String, isActivated: Boolean)`

There is a method to activate it like so

```
fun Repository.activateProduct(productId) {
    api.activateProduct(productId)
     .flatMap { db.updateProduct(productId, isActivated=true) }
}
```

ViewModel has a stream of Product objects as a database Observable

```
ViewModel {
   init {
      repository.productObservable
         .execute {
            copy(product = it)
         }
   }

   fun activate() {
      withState {
         repository.activateProduct(product.id)
            .execute {
               copy(activationAsnyc = it)
            }
         }
   }
}

data class State(product: Product, activationAsnyc: Async)
```

My issue with this is, that since `db.updateProduct` is writing to the database, which means, it will cause the `repository.productObservable` to emit - before Async is Success.

Tldr; there is point in time where `State(product.isActivated=true, ..)` while `Async` is still `Loading`
Is there a way around this or how did you dealt with this in your experience?

**Comment by gpeal** [comment]
@ursusursus What is the issue you're having with this as is? As long as your UI is reactive to the state, they should both show the correct values basically at the same time since both observables will emit after the same db transaction. 

Consider making derived properties in the body of your state class if you want to have new properties that are strictly derived from the main properties like product and activationAsync.

**Comment by ursusursus** [comment]
The issue is more of a academic one, but it bothers me since mvi-ish state modeling should be in a way where invalid states are unrepresentable.

Yes eventually it will be correct. However for few ms there will exist a state where State(product.activated=true, activationAsync=Loading) or State(product.activated=false, activationAsync=Success) depending on randomness of thread scheduling, (most likely the latter one) since product table will be written to frist, then function returns and exits the `activateProduct` which will emit activationAsync=Success synchronously. Now few ms later `productObservable` will emit since its on Schedulers.io.

Therefore viewmodel will see very quick 2 state changes, of which only the last one is valid. And I dont feel like using debounce and friends, since this is inhere to every database observables + async pair

**Comment by ursusursus** [comment]
Here is an example, its not mvrx just yet but its same concept

```
class MainViewModel(private val productStore: ProductStore) : BaseViewModel<MainViewModel.State>(State()) {

    override fun init() {
        disposables += productStore.productObservable()
            .observeOn(AndroidSchedulers.mainThread())
            .throwingSubscribe {
                setState { copy(product = it.value) }
            }
    }

    fun activateProduct() {
        val activate = !(state().product?.activated ?: false)

        disposables += productStore.activateProduct(activate)
            .subscribeOn(Schedulers.io())
            .asAsync(AndroidSchedulers.mainThread())
            .throwingSubscribe {
                setState { copy(activationAsync = it) }
            }
    }

    data class State(val product: Product? = null, val activationAsync: Async<Unit> = Uninitialized)
}
```
```
class ProductStore(private val productQueries: ProductQueries) {

    fun productObservable(): Observable<Optional<Product>> = productQueries.product().asObservable().mapToOptional()

    fun insertProduct() {
        productQueries.insertProduct(false)
    }

    fun activateProduct(activate: Boolean): Single<Unit> {
        return Single.just(Unit)
            .flatMap { Single.timer(1, TimeUnit.SECONDS).map { Unit } }
            .doOnSuccess { updateProductActivated(activate) }
    }

    fun updateProductActivated(activated: Boolean) {
        productQueries.updateProductActivated(activated)
    }
}
```

After ViewModel.activateProduct is called it outputs state
```
01-26 17:32:39.135 10684-10684/foo D/Default: state=State(product=Product.Impl [
      id: 1
      activated: false
    ], activationAsync=Loading@cfd300b)
01-26 17:32:40.158 10684-10684/foo D/Default: state=State(product=Product.Impl [
      id: 1
      activated: false
    ], activationAsync=Success(value=kotlin.Unit))
01-26 17:32:40.159 10684-10684/foo D/Default: state=State(product=Product.Impl [
      id: 1
      activated: true
    ], activationAsync=Success(value=kotlin.Unit))
```

As you can see the middle emit is exactly what I described above, and its invalid technically. Unless state is somehow debounced

**Comment by gpeal** [comment]
In practice, we have followed this pattern of eventual consistency and it works well. If you want to prevent certain states from being possible, you are welcome to do so however you please but MvRx doesn't enforce it by design.

I'm going to close this unless there is a specific bug or feature request here.
